google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITDMLTest.java

```diff
@@ -363,6 +416,214 @@ public void standardDMLAndMutations() {
+        .isNull();
+  }
+
+  @Test
+  public void testInsertOrUpdateOnNewRow() {
+    executeUpdate(DML_COUNT_WITH_UPSERT, insertOrUpdateDml(dialect.dialect));
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(10);
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo5", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(50);
+
+    // Delete the rows
+    executeUpdate(DML_COUNT_WITH_UPSERT, deleteDml());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V")))
+        .isNull();
+  }
+
+  @Test
+  public void testInsertOrIgnoreDmlOnExistingRow() {
+    // First insert few rows.
+    executeUpdate(DML_COUNT, insertDml());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(1);
+
+    // Existing rows are ignored, only 1 new row is inserted.
+    executeUpdate(1, insertOrIgnoreDml(dialect.dialect));
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(1);
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo5", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(5);
+
+    // Delete the rows
+    executeUpdate(DML_COUNT_WITH_UPSERT, deleteDml());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V")))
+        .isNull();
+  }
+
+  @Test
+  public void testInsertOrIgnoreDuplicateRows() {
+    long expectedRowCount = 2;
+
+    // Only unique keys are inserted. Duplicate keys are ignored.
+    if (dialect.dialect == Dialect.GOOGLE_STANDARD_SQL) {
+      executeUpdate(
+          expectedRowCount,
+          "INSERT OR IGNORE INTO T (k, v) VALUES ('1-boo1', 1), ('2-boo2', 2), ('1-boo1', 1)");
+    } else {
+      executeUpdate(
+          expectedRowCount,
+          "INSERT INTO T (k, v) VALUES ('1-boo1', 1), ('2-boo2', 2), , ('1-boo1', 1) "
```

Comment:
I see a syntax error in this line because of an extra comma: INSERT INTO T (k, v) VALUES ('1-boo1', 1), ('2-boo2', 2), **, (**'1-boo1', 1)
Pls remove the extra comma.
`INSERT INTO T (k, v) VALUES ('1-boo1', 1), ('2-boo2', 2), ('1-boo1', 1)`